src/test/java/org/commcare/formplayer/services/FormDefinitionServiceTest.java

```diff
@@ -162,20 +177,133 @@ public void testGetOrCreateFormDefinitionNewVersionNotEqual() {
+    }
+
+    @Test
+    public void testWriteToLocalStorage_FormDefExists() {
+        getFormDefStorage().write(formDef);
+        boolean valueWritten = this.formDefinitionService.writeToLocalStorage(this.formDef);
+        assertThat(valueWritten).isFalse();
+    }
+
+    @Test
+    public void testGetFormDefCaching() {
+        SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+        String sessionId = UUID.randomUUID().toString();
+        SerializableFormSession session = new SerializableFormSession(sessionId);
+        session.setFormDefinition(formDef);
+
+        assertThat(getCachedFormDefinition(sessionId)).isEmpty();
+        this.formDefinitionService.getFormDef(session);
+        assertThat(getCachedFormDefinition(sessionId)).isNotEmpty();
+    }
+
+    @Test
+    public void testGetFormDefBrokenSerialization() {
+        SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+
+        // put a bad value in `serializedFormDef` to trigger a serialization error
+        ReflectionTestUtils.setField(formDef, "serializedFormDef", "not a form def");
+        SerializableFormSession session = new SerializableFormSession(UUID.randomUUID().toString());
+        session.setFormDefinition(formDef);
+
+        // make sure we're not going to hit the cache
+        assertThat(getCachedFormDefinition(session.getId())).isEmpty();
+
+        // write the formDef to storage so that we can read it back
+        getFormDefStorage().write(this.formDef);
+        assertThat(this.formDef.getID()).isNotNull();
+
+        FormDef reSerializedFormDef = this.formDefinitionService.getFormDef(session);
+        assertThat(reSerializedFormDef.getID()).isEqualTo(this.formDef.getID());
+
+    }
+
+    @Test
+    public void testGetFormDefFromStorageUpdatesSerializableFormDefinition() throws Exception {
+        SerializableFormDefinition formDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+        ReflectionTestUtils.setField(formDef, "serializedFormDef", "not a form def");
+        SerializableFormSession session = new SerializableFormSession(UUID.randomUUID().toString());
+        session.setFormDefinition(formDef);
+        getFormDefStorage().write(this.formDef);
+
+        FormDef reSerializedFormDef = this.formDefinitionService.getFormDef(session);
+
+        SerializableFormDefinition updatedFormDef = this.formDefinitionService.getOrCreateFormDefinition(
+                this.appId, this.formXmlns, this.formVersion, this.formDef
+        );
+        assertEquals(updatedFormDef.getSerializedFormDef(),
+                FormDefStringSerializer.serialize(reSerializedFormDef));
```

Comment:
I don't think this is actually testing anything since the broken form def was never saved to the DB. I think you'd need to save the broken SerializableFormDefinition before calling `formDefinitionService.getFormDef(session)`.

You can confirm this by removing the call to `this.formDefinitionService.getFormDef(session)` and seeing if the test passes.

Reply:
Same goes for `testGetFormDefFromStorageUpdatesCache`